flink/v1.16/flink/src/main/java/org/apache/iceberg/flink/source/assigner/EventTimeAlignedAssigner.java

```diff
@@ -0,0 +1,301 @@
+
+public class EventTimeAlignedAssigner implements SplitAssigner, WatermarkTracker.Listener {
+  private static final Logger log = LoggerFactory.getLogger(EventTimeAlignedAssigner.class);
+  private final WatermarkTracker watermarkTracker;
+  private final State state;
+  private final TimestampAssigner<IcebergSourceSplit> timestampAssigner;
+  private final FutureNotifier futureNotifier;
+  private final Options options;
+
+  // flag that informs the state that there are no more splits to be added by the Enumerator
+  private volatile boolean noMoreSplits;
+
+  public EventTimeAlignedAssigner(
+      WatermarkTracker watermarkTracker,
+      TimestampAssigner<IcebergSourceSplit> timestampAssigner,
+      Options options) {
+    this.watermarkTracker = watermarkTracker;
+    this.state = new State(watermarkTracker, timestampAssigner);
+    this.options = options;
+    this.timestampAssigner = timestampAssigner;
+    this.futureNotifier = new FutureNotifier();
+    this.watermarkTracker.onInitialization();
+  }
+
+  @Override
+  public void start() {
+    watermarkTracker.addListener(this);
+  }
+
+  @Override
+  public void close() {
+    watermarkTracker.removeListener(this);
+    notifyListener();
+  }
+
+  @Override
+  public GetSplitResult getNext(@Nullable String hostname) {
+    GetSplitResult result = getNextInternal();
+    if (result.status() == GetSplitResult.Status.AVAILABLE) {
+      state.onSplitAssigned(result.split());
+    }
+    return result;
+  }
+
+  @Override
+  public void onDiscoveredSplits(Collection<IcebergSourceSplit> splits) {
+    state.onDiscoveredSplits(splits);
+  }
+
+  @Override
+  public void onUnassignedSplits(Collection<IcebergSourceSplit> splits) {
+    state.onUnassignedSplits(splits);
+  }
+
+  @Override
+  public void onCompletedSplits(Collection<String> completedSplitIds) {
+    state.onCompletedSplits(completedSplitIds);
+    if (!completedSplitIds.isEmpty() && isTerminal()) {
+      updateListenersOnTerminalCondition();
+    }
+  }
+
+  @Override
+  public void onNoMoreSplits() {
+    noMoreSplits = true;
+    if (isTerminal()) {
+      updateListenersOnTerminalCondition();
+    }
+  }
+
+  private boolean isWithinBounds(IcebergSourceSplit split, Long watermark) {
+    if (watermark == null) {
+      return true;
+    }
+
+    long splitTs = timestampAssigner.extractTimestamp(split, -1);
+    if (splitTs < watermark) {
+      log.warn("splitTs at {} is lower than the watermark {}", splitTs, watermark);
+    }
+
+    return Math.max(splitTs - watermark, 0L) <= options.getThresholdInMs();
+  }
+
+  private GetSplitResult getNextInternal() {
+    if (state.getUnassignedSplits().isEmpty()) {
+      log.info("looks like there are no splits to be assigned");
+      return GetSplitResult.unavailable();
+    }
+
+    try {
+      Long watermark = watermarkTracker.getGlobalWatermark();
+      IcebergSourceSplit pendingSplit = getFirst(state.getUnassignedSplits(), null);
```

Comment:
We can replace this line with:
```
IcebergSourceSplit pendingSplit = state.getUnassignedSplits()
    .stream()
    .filter(Objects::nonNull)
    .findFirst()
    .orElse(null);
```

This gets rid of using `org.apache.flink.calcite.shaded.com.google.common.collect.Iterables.getFirst`, so we can remove that import.